src/cargo/sources/registry/local.rs

```diff
@@ -70,38 +70,39 @@ impl<'cfg> RegistryData for LocalRegistry<'cfg> {
     }
 
     fn download(&mut self, pkg: PackageId, checksum: &str) -> CargoResult<MaybeLock> {
-        let crate_file = format!("{}-{}.crate", pkg.name(), pkg.version());
-        let mut crate_file = self.root.open_ro(&crate_file, self.config, "crate file")?;
-
-        // If we've already got an unpacked version of this crate, then skip the
-        // checksum below as it is in theory already verified.
-        let dst = format!("{}-{}", pkg.name(), pkg.version());
-        if self.src_path.join(dst).into_path_unlocked().exists() {
-            return Ok(MaybeLock::Ready(crate_file));
+        let filename = format!("{}-{}.crate", pkg.name(), pkg.version());
+
+        // Attempt to open an read-only lock first to avoid an exclusive write lock.
+        //
+        // If this fails then we fall through to the exclusive path where we copy
+        // the file.
+        if let Ok(dst) = self.cache_path.open_ro(&filename, self.config, &filename) {
+            let meta = dst.file().metadata()?;
+            if meta.len() > 0 {
+                return Ok(MaybeLock::Ready(dst));
+            }
```

Comment:
This is directly copied from `RemoteRegistry`. Probably not necessary, though I'm wondering how necessary it is in the current version of `RemoteRegistry` anyway.